docs/enhancements/aws-private-link.md

```diff
@@ -0,0 +1,553 @@
+$ SERVICE_AVAILABITY_ZONES=$(jq '.ServiceDetails[0].AvailabilityZones')
+```
+```shell=bash
+$ SUBNETS=$(subnetsForAZs SERVICE_AVAILABITY_ZONES)
+$ AWS_PROFILE=hub01 aws ec2 create-vpc-endpoint \
+--vpc-id <vpc-id> \
+--vpc-endpoint-type Interface \
+--service-name <service name> \
+--subnet-ids $SUBNETS \
+--security-group-ids <security group that allows ingress> \
+--private-dns-enabled
+```
+NOTE: To use private DNS, you must set the following VPC attributes 
+to true: enableDnsHostnames and enableDnsSupport.
+
+Now the `api.hash123.hub01osdprivate.io:6443` from the hub01 VPC should use the
+setup Endpoint Interface and Endpoint Service to reach the private API of customer's
+cluster.
+
+Update the kubeconfig used by the OpenShift installer to use new address.
+```shell=bash
+# modify the install_dir01/auth/kubeconfig
+# replace api.clustername.basedomain:6443 => api.hash123.hub01osdprivate.io:6443
+# replace the trust bundle to new trust
+```
+
+Now run the rest of the stages for the installer.
+```shell=bash
+$ openshift-install --dir install_dir01 wait-for bootstrap-complete
+$ openshift-install --dir install_dir01 destroy bootstrap
+$ openshift-install --dir install_dir01 wait-for install-complete
+# install complete
+```
+
+Update the ClusterDeployment's kubeconfig used for remote calls to use the
+new address and trust.
+
+### AWS Private Link Controller
+
+The goal of the controller to ensure a PrivateLink is established to the cluster
+that is being created by a ClusterDeployment. The controller watches for
+ClusterProvisioning and ClusterDeployments that require a PrivateLink and uses 
+the information in the aws platform to set it up.
+
+#### API change
+
+The ClusterDeployment needs to include additional information for the controller
+to create Privatelink. Some of the information that is required are,
+- infrastructure identifier for the cluster to discover the Network Load Balancer
+- private DNS name for the service
+- User/Role in AWS account (hub account) where the endpoint is to be created
+- subnets in AWS account (hub account) where the endpoint is to be created
+- credentials to the AWS account (hub account)
+
+```yaml=
+# clusterdeployment
+spec:
+  platform:
+    aws:
+      privatLink:
+        # this is the public r53 zone in hub account where the TXT records will be 
+        # added for domain verification of endpoint service
+        hostedZone: (string)
+        # this is list of subnets in hub account for the endpoint
+        # the subnets used for the endpoint depends on AZs supported by the
+        # service (which depends on the AZs supported by the NLB)
+        subnets: (list(string))
+        # this is reference to a secret that provide credentials to create
+        # necessary resources in the hub account.
+        # This role/user is only IAM entity that will be allowed to create endpoints
+        # to the service created for the cluster.
+        credentialsSecretRef:
+          name: (string)
```

Comment:
It feels like this batch of info belongs global on the HiveConfig, the team which creates our ClusterDeployments is unlikely to know this info and not be happy about having to look it up, whereas the team which configures a Hive installation (SRE/AppSre) would be a much better fit. Could we model this as global information in HiveConfig?

Reply:
Added to the alternatives to make sure we discuss and report the decision based on current design.

here's why i think per clusterdeployment makes sense..

the list of subnets depends on the region that was chosen for the cluster
the hostedZone depends on the apiOverrideURL chosen for the cluster
And the hive components need network access to the PrivateLink setup using the VPC Endpoint.

These requirements need knowledge of the environment (where hive is) and the cluster that is being created. There is no good way to match the environment values to the cluster values requested by the user unless hive understands the contents of the install config as i see today. so the simplest would be let user gives us the write group of these values.

as we get smarter to figure these out (i think we can) we can make these fields optional.

-----

Hive controller has 
- a list of suported HostestZones (domain names)
- a list of subnets in each supported region
- credentials to make things happen

Given those values we can
- restrict that all apiOverrideURL be subdomains on the hosted zone
- only specific regions be supported

When the controller has to create VPC Endpoint it looks at the region of the cluster, and finds those subnets to use.

Reply:
Did alternatives section not get pushed yet?